From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jaime Gonzalez <jaime.gonzalez@appentra.com>
Date: Mon, 13 Oct 2025 09:00:00 +0200
Subject: Allow parse macro names in statements and expressions

---
 grammar.js | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/grammar.js b/grammar.js
index b5c11d1..c6ffa3b 100644
--- a/grammar.js
+++ b/grammar.js
@@ -1182,6 +1182,7 @@ module.exports = grammar({
       // Not strictly valid, but can catch extensions and preprocessor macros
       $.call_expression,
       $.include_statement,
+      $.macro_identifier
     ),
 
     statement_label: $ => prec(1, alias($._integer_literal, 'statement_label')),
@@ -1945,6 +1946,7 @@ module.exports = grammar({
       $.implied_do_loop_expression,
       $.coarray_expression,
       $.conditional_expression,
+      $.macro_identifier
     ),
 
     parenthesized_expression: $ => seq(
